ee/clickhouse/models/property.py

```diff
@@ -32,12 +33,100 @@
+from posthog.models.property import (
+    NEGATED_OPERATORS,
+    OperatorType,
+    Property,
+    PropertyGroup,
+    PropertyIdentifier,
+    PropertyName,
+)
 from posthog.models.team import Team
 from posthog.utils import is_valid_regex, relative_date_parse
 
+# Property Groups Example:
+# {type: 'AND', groups: [
+#     {type: 'OR', groups: [A, B, C]},
+#     {type: 'OR', groups: [D, E, F]},
+# ]}
+
+# Example:
+# {type: 'AND', groups: [
+#     A, B, C, D
+# ]}
+
+# Property json is of the form:
+# { type: 'AND | OR', groups: List[Property] }
+# which is parsed and sent to this function ->
+
+
+def parse_prop_grouped_clauses(
+    property_group: PropertyGroup,
+    prepend: str = "global",
+    table_name: str = "",
+    allow_denormalized_props: bool = True,
+    has_person_id_joined: bool = True,
+    person_properties_mode: PersonPropertiesMode = PersonPropertiesMode.USING_SUBQUERY,
+    person_id_joined_alias: str = "person_id",
+    group_properties_joined: bool = True,
+    _top_level: bool = True,
+) -> Tuple[str, Dict]:
+
+    if len(property_group.groups) == 0:
+        return "", {}
+
+    if isinstance(property_group.groups[0], PropertyGroup):
+        group_clauses = []
+        final_params = {}
+        for idx, group in enumerate(property_group.groups):
+            if isinstance(group, PropertyGroup):
+                clause, params = parse_prop_grouped_clauses(
+                    property_group=group,
+                    prepend=f"{prepend}_{idx}",
+                    table_name=table_name,
+                    allow_denormalized_props=allow_denormalized_props,
+                    has_person_id_joined=has_person_id_joined,
+                    person_properties_mode=person_properties_mode,
+                    person_id_joined_alias=person_id_joined_alias,
+                    group_properties_joined=group_properties_joined,
+                    _top_level=False,
+                )
+                group_clauses.append(clause)
+                final_params.update(params)
+
+        _final = f"{property_group.type} ".join(group_clauses)
+    else:
+        _final, final_params = parse_prop_clauses(
+            filters=cast(List[Property], property_group.groups),
+            prepend=f"{prepend}",
+            table_name=table_name,
+            allow_denormalized_props=allow_denormalized_props,
+            has_person_id_joined=has_person_id_joined,
+            person_properties_mode=person_properties_mode,
+            person_id_joined_alias=person_id_joined_alias,
+            group_properties_joined=group_properties_joined,
+            property_operator=property_group.type,
+        )
+
+    if not _final:
+        final = ""
+    elif _top_level:
+        final = f"AND ({_final})"
+    else:
+        final = f"({_final})"
+
+    return final, final_params
+
+
+def is_property_group(group: Union[Property, "PropertyGroup"]):
```

Comment:
This is a smell in that PropertyGroup just wants to be another kind of Property rather than it's own entity.

I suggested this in my initial review - not sure why a different approach was chosen but I strongly disagree with it.